set.mm

```diff
@@ -22442,40 +22506,32 @@ of the unique existential quantifier (note that ` y ` and ` z ` need not
   ${
     $d x y z $.  $d ph z $.
     euf.1 $e |- F/ y ph $.
-    $( A version of the existential uniqueness definition with a hypothesis
-       instead of a distinct variable condition.  (Contributed by NM,
-       12-Aug-1993.)  (Proof shortened by Wolf Lammen, 30-Oct-2018.) $)
+    $( Version of ~ eu6 with DV condition replaced by nonfreeness hypothesis.
+       (Contributed by NM, 12-Aug-1993.)  (Proof shortened by Wolf Lammen,
+       30-Oct-2018.) $)
     euf $p |- ( E! x ph <-> E. y A. x ( ph <-> x = y ) ) $=
       ( vz weu weq wal wex eu6 nfv nfbi nfal equequ2 bibi2d albidv cbvex bitri
       wb ) ABFABEGZSZBHZEIABCGZSZBHZCIABEJUBUEECUACBATCDTCKLMUEEKECGZUAUDBUFTUC
       AECBNOPQR $.
   $}
 
-  ${
-    $d x y z $.  $d ph z $.
-    mo2.1 $e |- F/ y ph $.
-    $( Alternate definition of the at-most-one quantifier.  (Contributed by NM,
-       8-Mar-1995.)  Extract ~ mo2v from this proof, and prove ~ mo2 from it.
-       (Revised by Wolf Lammen, 28-May-2019.) $)
-    mo2 $p |- ( E* x ph <-> E. y A. x ( ph -> x = y ) ) $=
-      ( vz wmo weq wi wal wex df-mo nfim nfal equequ2 imbi2d albidv cbvex bitri
-      nfv ) ABFABEGZHZBIZEJABCGZHZBIZCJABEKUBUEECUACBATCDTCSLMUEESECGZUAUDBUFTU
-      CAECBNOPQR $.
-  $}
-
   ${
     $d x y $.  $d y ph $.
-    $( Bound-variable hypothesis builder for uniqueness.  (Contributed by NM,
-       9-Jul-1994.)  (Revised by Mario Carneiro, 7-Oct-2016.) $)
-    nfeu1 $p |- F/ x E! x ph $=
+    $( Alternate proof of ~ nfeu1 .  Shorter but using ~ eu6 and a dummy
+       variable.  Note that ~ nfeu1 has no usage before ~ nfeu1ALT so that the
+       latter could be used, but ~ nfeu1 's proof is more systematic.
+       (Contributed by NM, 9-Jul-1994.)  (Revised by Mario Carneiro,
+       7-Oct-2016.)  (Proof modification is discouraged.)
+       (New usage is discouraged.) $)
+    nfeu1ALT $p |- F/ x E! x ph $=
```

Comment:
In what way is this shorter?

Reply:
```
---------The proof of "nfeu1" (74 bytes) ends above this line.
---------The proof of "nfeu1ALT" (73 bytes) ends above this line.
```

Reply:
It seems nfeu1ALT is shorter, because it has 4 lines, and nfeu1 has 5.  This makes nfeu1ALT be the preferred candidate.

Reply:
Yes, I wrote in its comment that it is shorter, and I also wrote why nfeu1 is preferred (more systematic, no dummy variable, and nfeu1ALT uses eu6 while nefeu1 is proved directly from the definition):
```
sh pr nfeu1
14   nfbii.1=df-eu $a |- ( E! x ph <-> ( E. x ph /\ E* x ph ) )
20     nfan.1=nfe1   $p |- F/ x E. x ph
23     nfan.2=nfmo1  $p |- F/ x E* x ph
24   nfxfr.2=nfan  $p |- F/ x ( E. x ph /\ E* x ph )
25 nfeu1=nfxfr   $p |- F/ x E! x ph
sh pr nfeu1ALT
17   nfbii.1=eu6    $p |- ( E! x ph <-> E. y A. x ( ph <-> x = y ) )
23     nfex.1=nfa1    $p |- F/ x A. x ( ph <-> x = y )
24   nfxfr.2=nfex   $p |- F/ x E. y A. x ( ph <-> x = y )
25 nfeu1ALT=nfxfr $p |- F/ x E! x ph
```

Reply:
From conventions:

If two proofs have the same number of compressed proof steps, we will typically give preference to the one with the smaller number of different labels

which in our case is nfeu1ALT.  It is of no importance whether it declares a dummy variable.  Shorter proofs also make better use of other theorems, so the use of eu6 is no downside either.  And the idea of systematic is disputable, too.  And is no criterium either.

Reply:
I suppose we should use the shorter version although I'm having trouble forming opinions either way.

Reply:
I know it does not follow the convention, and that is the reason I wrote these explanations in the comment of nfeu1ALT. As for "systematic", it is precise here in the sense that nfeu1 follows the standard algorithm to prove nonfreeness in a defined expression (just regard the definiens as a tree whose nodes are formulas and prove by tree-induction nonfreeness at each node, from the leaves up to the root). Let's see if we can get an opinion from the originator of the nonfreeness predicate, @digama0. If we cannot (or obviously if he agrees with you), then I'll swap nfeu1 and nfeu1ALT.

Reply:
I don't know what to say here.  Maybe this: First, I think, you need to provide some material why the conventions fail - not in your case alone, but on a broader scale.  The material should convince the community that the conventions need an update.  Once they allow your move, then it will be fine with me.

Reply:
I did provide a reason why this should be an exception, both in the initial comment of ~nfeu1ALT, and then with further precisions in the above comment.  This did not convince you, so I just swapped nfeu1 and nfeu1ALT in the latest commit, with appropriate comments.  I hope that's ok now.

Note that I had to move nfeu1(ALT) after eu6. That is, we first prove an equivalent definition (eu6) and from it we prove nonfreeness.  In the future, it'd be nice to automate the statement and proof of the "staple theorems" associated with each new definition (e.g., df-xxx is followed by the equality theorems like xxxeq1, xxxeq2, xxxbi1... the nonfreeness theorems when they apply, nfxxx1/nfxxx, the existence theorem when it applies, xxxeq, etc.).  I think that @digama0 implemented some of this in mmj2 (and probably plans it for MM0 ?). With this automation, the resulting proof of nfeu1 would be the one I proposed.

Reply:
> I just swapped nfeu1 and nfeu1ALT in the latest commit, with appropriate comments. I hope that's ok now.

Thanks. Like I said before, I don't really have an opinion one way or the other, but this seems like the expedient thing at least for now.

> In the future, it'd be nice to automate the statement and proof of the "staple theorems" associated with each new definition (e.g., df-xxx is followed by the equality theorems like xxxeq1, xxxeq2, xxxbi1... the nonfreeness theorems when they apply, nfxxx1/nfxxx, the existence theorem when it applies, xxxeq, etc.). I think that @digama0 implemented some of this in mmj2 (and probably plans it for MM0 ?). With this automation, the resulting proof of nfeu1 would be the one I proposed.

Oooh, interesting. This is getting my mind running about all the ways definitions could (hypothetically) be better (especially in a next generation system like ghilbert, MM0, or whatever it ends up being). But I suppose coming up with ideas has been the easy part of those efforts, but writing code and dealing with the annoying details has been the hard part. I guess I'll go back to dreaming at least for now.

Reply:
As a matter of fact, MM0 does generate equality theorems automatically for every definition, and it could generate nf theorems as well but it wasn't too important since I don't have many fancy binders that would need it.